### ISSUE TITLE
Fix repeated block timeouts during initial sync

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#5709]
## Motivation

During the initial sync, Zebra's block validation repeatedly times out, and the syncer resets:
> 2022-11-23T09:22:15.197732Z  WARN {net="Main"}:sync:try_to_sync: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1856228), hash: block::Hash("0000000000d433e30155445ddc72854eb7395827c1c241b249a255585ea2716a") }

> https://github.com/ZcashFoundation/zebra/actions/runs/3519147385/jobs/5923785413#step:8:12498

This only happens in some parts of the chain, and it seems to happen at slightly different block heights each time. When it does happen, it can happen repeatedly every 10-15 minutes:
https://github.com/ZcashFoundation/zebra/actions/runs/3519147385/jobs/5923785413#step:8:12465

This is not a priority because Zebra is still much faster than `zcashd`.

### Designs

This could be happening because:
- blocks are committed out of order
- we're downloading blocks in the wrong order (not strict height order) then hitting a concurrency limit so we can't download the rest before the timeout
- the block timeout is too short for the number of blocks allowed in the validate and commit pipeline
  - decreasing the pipeline capacity also reduces RAM use, but makes runs of small blocks download and verify slower
  - increasing the timeout increases the amount of time we wait for blocks that will never verify, increasing RAM usage
  - we tried increasing the timeout and it didn't work, but a slight increase might be ok
- our state isn't responding to all the UTXOs for the blocks it has (I think this is a known issue)
- the verifier is blocked in some other way

## Related Work

If we speed up block commits, it might solve this issue:
- #5425

**Comment by teor2345** [comment]
Due to this block verification timeout bug, full syncs are currently taking 40-48 hours. That's still acceptable performance, I think?

**Comment by mpguerra** [comment]
Is this still happening? if so, when was the last time we saw it?

**Comment by teor2345** [comment]
> Is this still happening? if so, when was the last time we saw it?

Yes, it happens multiple times in every full sync, this sync is running right now:
https://github.com/ZcashFoundation/zebra/actions/runs/4931890966/jobs/8824296617#step:8:8745

In the last completed full sync, it happened at least 200 times, which causes a delay of 3-5 hours in  50 hour sync:
https://github.com/ZcashFoundation/zebra/actions/runs/4888643906/jobs/8769200765#step:8:17512

**Comment by mpguerra** [comment]
@ZcashFoundation/zebra-team has anyone noticed this happening recently? I'm tempted to close

**Comment by upbqdn** [comment]
Yes, we still have this bug. I got a few of the errors while running an unsynced instance locally:

```
2024-03-13T13:23:23.783618Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1941142), hash: block::Hash("0000000000aca7473124dd16884dfe35ee602ba7f9dda66bc2633563c5db3a4e") }
2024-03-13T13:34:23.516921Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1942686), hash: block::Hash("0000000000ffb91722514ad09b493355add5b8f5559600b7dd7c7b715cdc80a2") }
2024-03-13T13:44:11.547760Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1944827), hash: block::Hash("0000000001b3c62ba33f778464b308a0435f25a2afa995d1bbbfa553e29c9878") }
2024-03-13T15:31:57.838278Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1977421), hash: block::Hash("00000000016daccf7867a7b7ad930e118d5f5cd2b13612eea2df038c7aa9747b") }
2024-03-13T15:41:43.896349Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1978958), hash: block::Hash("00000000006886ebe06217ded4b4718af82549afab6a1a66354c34cec93882be") }
2024-03-13T15:51:26.100828Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1980798), hash: block::Hash("00000000016422ba324158bd0ed9597dbcdec3a35ee94f08968a92b2440e1b9a") }
2024-03-13T16:01:18.321517Z  WARN {zebrad="58bfe97" net="Main"}:sync:try_to_sync:try_to_sync_once{extra_hashes={}}: zebrad::components::sync: error downloading and verifying block e=ValidationRequestError { error: Elapsed(()), height: Height(1982826), hash: block::Hash("0000000000db2bf60283410f90846223e981d3a4c0d5d9f656804be36067d163") }
```

The timing is interesting and doesn't seem random. It looks like every few hours, we get a series of error logs that are ten minutes apart.

**Comment by MDrollette** [comment]
First time trying to sync (v2.2.0) and I'm running into this. I wasn't paying attention to the logs at first but the weird network graph made me investigate. Clearly happening at a regular interval.

<img width="1047" alt="Image" src="https://github.com/user-attachments/assets/bed49d82-bb67-4aa5-84c0-8c679e5622d8" />

**Comment by mpguerra** [comment]
Let's look into this one after we're done with the zcashd deprecation milestone in a couple of sprints time

**Comment by upbqdn** [comment]
@MDrollette, could you report how long the sync took?

**Comment by MDrollette** [comment]
It didn't seem to be making any progress - the sync progress was showing 62-64% for a few days so I just shut it down. But I didn't look too closely at it to be honest since I was only running the node on a whim. I'll try to sync it again and get some more concrete times and logs if this seems like a real issue and not just a me-problem.

**Comment by MDrollette** [comment]
I started again from scratch and fully synced in around 60 hours.

It took about 7 hours to reach the point where warnings started (59%).
The slow period with warnings every 10 minutes lasted for approximately 45 hours (from 59% to 70%).
After the warnings stopped (at 70%), it took another 11 hours to reach 100%.

**Comment by upbqdn** [comment]
Great, thank you.

**Comment by pktpls** [comment]
Have stumbled upon this too the past few days. I didn't take notes so the numbers are a bit blurry. This is with the default config, I only played with the `sync` settings to see if it's a concurrency issue (doesn't look like it).

At around 60% it slows down a lot and prints ValidationRequestErrors, while doing seemingly nothing for minutes at a time, then sync normally for 20 seconds or a minute, pause again, repeat. During these pauses, inbound traffic drops to basically nothing, while outbound traffic is still going at full throttle (>300mbps).

With these pauses, Zebra seems substantially slower than Zcashd for this curious 60-70% part. Zcashd did take patience (between 12 and 24 hours) and then blasted through the last 30% in under 3 hours. I don't know yet what Zebra will do at 70%, I'm only at 61% so far.

So I took that synced Zcashd node and told Zebra to connect only to that (`peerset_initial_target_size = 1` & `crawl_new_peer_interval = "1d"` & `network.initial_mainnet_peers = [myzcashd:8233]`). Zebra's only connections are 3x to my Zcashd, but same pausing behaviour :-(

There are moments where Zebra and Zcashd exchange >1gbps, so the potential is there, but the pauses are just brutal 😅 The general slowdown between 60% and 70% seems unavoidable, but the pauses make things much worse and might be fixable I guess?

**Comment by mpguerra** [comment]
We're going to focus on investigating this issue in a couple of weeks time, hopefully during Sprint 11.

**Comment by mpguerra** [comment]
Could this be related to https://github.com/ZcashFoundation/zebra/issues/3438 and https://github.com/ZcashFoundation/zebra/discussions/3672 ?

**Comment by pktpls** [comment]
I just tried a fresh sync with zebrad 4.3.0 and it seems to be a lot better now. I don't see pauses in processing anymore, and no ValidationRequestErrors either. It did slow down a little between 59% and 65%, but now it's really blasting (.5% per minute).

**Comment by pktpls** [comment]
I jinxed it - apparantly right after my comment, at ~71% its networking collapsed. Every 5 to 15 minutes it manages to download a few hundred blocks, then back to doing nothing. Restarts of zebrad don't help.

Might be a different issue, I dont know. I'll keep it running for a few more hours.

**Comment by pktpls** [comment]
Okay it does seem like the same or very similar issue as before, I'm just puzzled why it's happening later in the sync now. (I'm around 2440000 at the moment.)

Edit: I am now at 2500000, there are short periods where it blasts, and then minutes where it's only uploading but not doing anything else.

**Comment by arya2** [comment]
This is likely due to block downloads failing after all retries breaking the contiguous chain of blocks in zebra-state's QueuedBlocks that can be sent to be contextually validated and committed.

**Comment by pktpls** [comment]
I played around with this a little more, it did sometimes hang for >10 minutes, but always recovered to reasonable speeds.

All in all, this sync took about 9 hours, which seems pretty good to me. Had I just left it running and not looked at the logs, I wouldn't have noticed a thing :-)

(Ryzen 9900X 12C/24T and 500/500 Mbps fiber)

**Comment by iBlackSunday** [comment]
Having this same issue. I was able to restrict the individual nodes that I was downloading from, as soon as I added specific node IP's, the node sync would get stuck waiting for peers to respond. Let this sit and sync for 3 days, was never able to get past 20%. Going to have to only enable these two Node IP to ensure chain sync, but this is a major issue and needs to be looked into more.

Working Node IP:
35.172.185.187
18.27.125.103

**Comment by zmanian** [comment]
### Reproducible on 4.5.1, with evidence narrowing the root cause to verify-stage parking (not download timeouts)

We hit this deterministically during initial sync on a cloud node and gathered evidence that I think discriminates between the hypotheses in the issue body.

**Environment**
- `zfnd/zebra:4.5.1` (also reproduced identically on 4.4.1), mainnet, Docker bridge.
- GCP c4-standard-24 (24 vCPU, 88 GB), hyperdisk-balanced 15K IOPS. Hardware was bumped 3x (c3-8 → c4-16 → c4-24) and storage swapped (pd-ssd → hyperdisk) with no effect.
- Stall region observed ~1.43-1.45M (checkpoint phase, far below the last checkpoint).

**Deterministic repro**
1. Start with a cleared peer cache. Watch `extending tips` walk `in_flight` up: 499 → 997 → 999.
2. Within ~90s of `in_flight` reaching ~999, the syncer goes **completely silent** — no `WARN`, no retries, no `trace`. Only the once-per-minute `estimated progress` line (separate task) keeps printing, with `time_since_last_state_block` climbing monotonically.
3. It self-recovers after ~8 minutes (`DownloadFailed { error: Elapsed(()) }` → restart), or sooner if we bounce the process. Each restart yields a 30-90s burst (~1,600-8,000 blocks, burst rate ~70-150K blocks/hr) then re-stalls.

**The discriminating evidence: blocks are parked in *verify*, not stuck in *download*.** During the silent window:
- Container CPU 0.01-0.07%, disk %util <1% (queue depth ~0).
- Container rx/tx at keepalive levels only (hundreds of B/s).
- **All ~30 ESTAB peer connections show 0 in both Recv-Q and Send-Q** — i.e. the node has **no outstanding block requests to any peer**. The 999 in-flight items are not waiting on the network; they have already downloaded.
- The silence duration matches `BLOCK_VERIFY_TIMEOUT = 8 * 60` (`sync.rs`), wrapping the verifier (`Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT)`) and `try_to_sync_once` (`timeout(BLOCK_VERIFY_TIMEOUT, ...)`).

This rules out the "block timeout too short to download the pipeline" branch (downloads completed; a genuine download straggler fails in ~60-80s via `BLOCK_DOWNLOAD_TIMEOUT` × retries and triggers a fast restart, which we also see separately). It points squarely at the **out-of-order download + strict-contiguity verifier** branch.

**Mechanism (as we read it):** `CheckpointVerifier::target_checkpoint_height()` walks upward from the last verified height and `break`s on the first gap, returning `WaitingForBlocks`; every buffered block above the gap keeps its response `oneshot` unresolved, so each is a `download_and_verify` task still counted in `in_flight`. Concurrently the downloader sets `past_lookahead_limit = true`, so the syncer's pause loop engages at `in_flight >= limit/2` and parks on `self.downloads.next().await` — it stops requesting blocks and stops walking tips, so it **cannot re-request the one missing/late block** in the lowest range. Nothing resolves until the 8-minute verify timeout forces a full restart. Restart re-walks from the (still-frozen) state tip, fills the gap bottom-up, and bursts until the next straggler creates a new gap.

**Note on tuning, in case it helps others:** `in_flight` is bounded by `checkpoint_verify_concurrency_limit` (default 1000), **not** `download_concurrency_limit`. We tested `download_concurrency_limit` 25 vs 50 — `in_flight` still saturated ~999, no change. We have not yet completed a controlled run with `checkpoint_verify_concurrency_limit` lowered toward its 400 floor (≈ one checkpoint gap), but in principle that should shrink the per-stall blast radius (fewer orphaned blocks above the gap) at the cost of pipeline depth, consistent with the "decreasing pipeline capacity" note in the issue body. Happy to report numbers if useful.

**Operationally** we are front-running the 8-minute timeout with a watchdog that restarts when `time_since_last_state_block` exceeds ~120s, which keeps the node ratcheting forward. It works but is obviously a workaround. The thing that would help most as a first step is observability: a `WARN` when `in_flight` is saturated and the state tip hasn't advanced for N seconds, ideally naming the contiguity gap height (the verifier already computes `gap` in a `trace!`) and the oldest in-flight block, so operators can see the wedge without attaching a debugger.

(Investigation assisted by Claude Code; all claims verified against the v4.4.1 and v4.5.1 source.)

**Comment by zmanian** [comment]
### Follow-up: even before the root cause is fixed, the stall should not be silent or unbounded

Separating the operability concerns from the root-cause discussion above, here are three small, independent improvements that would help operators regardless of how/when the contiguity bug itself is fixed. None of these is the fix — they make the wedge **observable** and **bounded**.

**1. WARN when saturated but the tip is not advancing.** Today the syncer parks at `self.downloads.next().await` in the `try_to_sync_once` pause loop and emits nothing at INFO for up to `BLOCK_VERIFY_TIMEOUT` (8 min); the only output is the once-per-minute `estimated progress` line from a separate task, so the node looks healthy-but-idle. A rate-limited `WARN` when `in_flight` is at/over the pause threshold **and** the state tip hasn't advanced for N seconds — including the contiguity **gap height** the verifier is waiting on (`CheckpointVerifier::target_checkpoint_height()` already computes `gap` in a `trace!`), the **oldest in-flight** block hash/height, plus `in_flight`/`lookahead_limit`/`state_tip` — would turn a silent 8-minute wedge into one actionable log line.

**2. Make stall recovery configurable or built-in.** The only escape from a wedged pipeline is the 8-minute `BLOCK_VERIFY_TIMEOUT`, which is far longer than a clean restart (process bounce ~10-30s + a 30-90s burst). Operators are currently reduced to an external watchdog that restarts on `time_since_last_state_block > ~120s`. Either making `BLOCK_VERIFY_TIMEOUT` (and optionally `SYNC_RESTART_DELAY`) configurable, or adding a built-in stall watchdog (e.g. `sync.force_restart_on_stall` / a stall timeout that triggers `downloads.cancel_all()` + restart when the tip is frozen while `in_flight` is saturated), would remove the need for an out-of-band bouncer and let operators trade recovery latency against wasted re-downloads explicitly.

**3. Document that `checkpoint_verify_concurrency_limit` is the real ceiling on `in_flight`.** It's bounded by `checkpoint_verify_concurrency_limit` (default 1000), not `download_concurrency_limit` — non-obvious, and it actively misleads tuning (lowering `download_concurrency_limit` leaves `in_flight` at ~999). A short doc note on both fields and on `lookahead_limit` would save operators time.

Happy to open these as a separate tracking issue if you'd prefer they not live inside #5709.

(Drafted with Claude Code; verified against v4.4.1 and v4.5.1 source.)

**Comment by zmanian** [comment]
One cross-link for whoever picks this up: a practical operator mitigation here is pinning known-good peers (a better peer set produces fewer late/missing blocks, which is what triggers the contiguity wedge described above). In env-var-only container/cloud deploys that workaround is currently blocked by **#10658** — collection-typed config fields like `initial_mainnet_peers` can't be set via `ZEBRA_*` env vars and require a mounted TOML file.

**Comment by zmanian** [comment]
### Proposed fix approach — looking for a maintainer steer before I write a PR

Following on from the diagnosis above, I've worked up a design and would like a maintainer's read on the **direction** before investing in a full PR (and on the one default-behavior change it implies).

**Framing constraint.** A literal "re-request the missing block by height" turns out not to fit the current architecture: the syncer is hash-keyed (`download_and_verify(hash)`, `cancel_handles: HashMap<Hash, _>`), heights are only known post-download, and a contiguity gap means the missing height's *hash* is unknown (it was last-hash-stripped in `obtain_tips`/`extend_tips`, or diverged across the fanout). So obtaining it requires re-walking tips via `FindBlocks` from the committed state tip. In other words, any real re-request reduces to "re-walk from the state tip and re-queue" — which is exactly what a manual restart does, and why a manual bounce reliably produces a burst. The bug is that the syncer only does this after the 8-minute `BLOCK_VERIFY_TIMEOUT` + 67s `SYNC_RESTART_DELAY`, because once saturated it parks on `downloads.next().await` and stops extending tips.

**Proposed design (no change to block validation/commit):**

1. `CheckpointVerifier` publishes a `watch::Sender<Option<block::Height>>` gap signal: `Some(contiguous_height)` while it is in `WaitingForBlocks` above a gap, `None` otherwise. The receiver is returned from the verifier-router init and threaded to the syncer. This is the only cross-crate surface change.

2. In the `try_to_sync_once` pause loop, race `downloads.next()` against a stall deadline. Trigger a **fast restart** only on a *persistent gap stall*: state tip frozen AND `in_flight` saturated AND `gap_signal == Some(_)` unchanged for `stall_restart_timeout`, with a ~30s thrash guard. The `gap_signal == None` case (a complete range verifying slowly) deliberately does **not** trigger, so we never throw away legitimate progress.

3. On that signal, `cancel_all()` + re-enter `obtain_tips()` immediately (short stall-restart delay instead of 67s; the 8-minute timeout remains a backstop). Because the pipeline is now empty, the low gap blocks are re-walked bottom-up before re-saturation.

4. Byproduct: a single rate-limited `WARN` naming the gap height, replacing the silent multi-minute wedge.

**The decision I'd most like input on:** I've speced this **on by default** (`sync.stall_restart_timeout = Some(90s)`, `None` to opt out), on the reasoning that the fix only helps users if it's not behind an undiscovered knob — but that does change default sync behavior, so it's your call. I'm equally happy to ship it opt-out/off-by-default if you'd prefer to be conservative.

Would you accept this direction? Any concerns with the verifier exposing a gap watch, or with the on-by-default default? If the approach is acceptable I'll prepare a PR with tests (TDD: a syncer stall→fast-restart test under paused tokio time, a verifier gap-signal test, and a disabled-path test). I have a full design doc written up and can share it or open a draft PR, whichever is easier to review.

(Design and investigation AI-assisted with Claude Code; I'm the responsible author and will disclose in the PR.)

**Comment by zmanian** [comment]
Adding a field report from a mainnet initial-sync stall we hit this week on a fresh GCP host running `zfnd/zebra:4.5.1`. Exactly matches the pattern described here — `in_flight` saturating, syncer task going silent, ~8-minute recovery cycle.

Two things worth noting for other operators landing here:

1. **`ZEBRA_SYNC__CHECKPOINT_VERIFY_CONCURRENCY_LIMIT=400` (the minimum) substantially reduces blast radius.** With the default 1000, `in_flight` saturates at 999 and each stall costs the full ~8 minutes. At 400, `in_flight` caps at 399 and we see sustained 12-18K blocks/minute through this region with `time_since_last_state_block=0s` and CPU at ~90%. Empirically much better than the two knobs whose names suggest they'd help (`download_concurrency_limit`, `peerset_initial_target_size`) — we tested both directions on those and `in_flight` was unchanged.

2. **The syncer task emits nothing during the stall.** No WARN, no peer-eviction event, no oldest-pending-request log. The only signal is `time_since_last_state_block` climbing in the once-per-minute `estimated progress` line. A WARN with the peer ID and block hash of the oldest in-flight request whenever the queue is saturated for >N seconds would collapse diagnostic time for operators dramatically.

Full write-up with diagnostic walk, things-that-don't-work list, final config, and a small bounce-watchdog as a safety net: https://gist.github.com/zmanian/08280e428b3f9b90551a2fa74a4b1a40

Thanks to everyone who's debugged this — the recommendation to clamp `checkpoint_verify_concurrency_limit` was the unlock.

**Comment by zmanian** [comment]
### A complete implementation is ready on a fork branch, if useful for evaluating the approach

To make the proposed direction concrete (and in case seeing the diff is easier than discussing it in the abstract), I've implemented it on a fork branch:

**https://github.com/zmanian/zebra/tree/fix/5709-sync-stall-pr** (based on v4.5.1)

What it contains:
- `CheckpointVerifier` publishes an observability-only `watch::Sender<Option<Height>>` gap signal (no change to verification/commit logic), threaded through `router::init` into the syncer.
- A `sync.stall_restart_timeout` config (`Duration`, default `90s`, `"0s"` disables) with a startup clamp below `BLOCK_VERIFY_TIMEOUT`.
- A `SyncError { Stalled, Other }` discriminant; the pause loop in `try_to_sync_once` races `timeout_at(last_tip_advance + timeout, downloads.next())` and, on a *persistent* gap (saturated queue + frozen tip + verifier gap signal `Some` and unchanged), returns `SyncError::Stalled`, which `sync()` fast-restarts from the current tip (re-walking tips bottom-up). A `WARN` names the gap height.
- Tests: pure-function branch coverage for the decision, a `#[tokio::test(start_paused = true)]` integration test driving the real loop to `Stalled`, a changed-then-persistent regression, and a deterministic verifier gap-signal test. `cargo build`, `clippy -D warnings`, `fmt`, and the sync + consensus suites are green.

It's intentionally minimal: it does not touch block validation, does not remove the last-hash-stripping behavior, and keeps the 8-minute `BLOCK_VERIFY_TIMEOUT` as a backstop. The default-on behavior change is the main thing I'd want your read on (happy to make it opt-out instead).

I'm deliberately **not** opening a PR yet — I'd rather get a maintainer's steer on whether this direction is acceptable first. If it is, I'll open it against this branch with the usual PR metadata and AI-tool disclosure. If you'd prefer a different approach (or this isn't a priority), no problem — just let me know.

(Implementation AI-assisted with Claude Code; reviewed before publishing. I'm the responsible author.)

**Comment by oxarbitrage** [comment]
@zmanian — good report, and it lines up closely with where our own analysis already landed, so before you open a PR: we've got a fix in progress and it'll be up shortly.

One of our leading theories was already on the block-verification side — the concurrency limit that controls how many blocks sit in the pipeline — and your report is what let us cleanly rule out the download and peer side. It confirms the mechanism: the checkpoint verifier stops at the first missing block, the ones above it pile up and saturate the pipeline, and the syncer then sits idle until the 8-minute timeout instead of re-requesting the block it's stuck on. Your 0 Recv-Q/Send-Q observation — no outstanding requests, parked in *verify*, not *download* — is exactly what made that call.

We've tried a few approaches over the last few days, none a clean fix on its own. So what we're shipping first is deliberately narrow, and it doesn't try to stop the gap from forming: it makes the node **recover from the wedge quickly** — a fast restart when the pipeline is saturated and the tip is frozen, instead of the silent 8-minute hang. No change to validation/commit. Why a given block goes missing in the first place is a separate, less settled question we're still working.

The gap-signal design you sketched is basically the same fix — ours is just a more minimal take on it. Once the PR is up, the paused-time TDD test you described (induce a stall → assert fast-restart under `tokio::time`) would be a genuinely useful piece to land on it together; I'll link you when it's open.

**Comment by zmanian** [comment]
Sounds great, and thanks — happy to defer to your fix; I won't open a PR. Glad the field data (the 0 Recv-Q/Send-Q "parked in verify, not download" observation) was useful in ruling out the download/peer side.

I'll watch for your PR and bring the paused-time stall→fast-restart test — induce a saturated-queue + frozen-tip stall under `tokio::time`, assert the syncer fast-restarts instead of hanging for the full verify timeout — plus the "gap changed once, then persistent" regression, adapted to your structure so it lands cleanly alongside. Just ping me when it's open.

**Comment by zmanian** [comment]
Follow-up data from running the experimental fix branch on real-world stalls.

## Setup
Running [zmanian/zebra@fix/5709-sync-stall](https://github.com/zmanian/zebra/tree/fix/5709-sync-stall) on a mainnet initial sync, GCP c4-standard-24 + hyperdisk-balanced. Configured with `stall_restart_timeout=90s`, `checkpoint_verify_concurrency_limit=400`, pinned peers (40 IPs + 4 DNS seeds), and Prometheus metrics enabled.

## What works
The new detector fires as designed when the original symptom appears:

```
2026-05-30T12:58:22  WARN sync:try_to_sync:try_to_sync_once: zebrad::components::sync:
syncer stalled on checkpoint gap; restarting from tip
gap_height=Some(1732097) state_tip=Some(Height(1732086)) in_flight=663 lookahead_limit=400
```

The 90s restart vs the 8-minute internal verifier timeout is a big win for those wedges.

## What the fix misses

After running for ~90 minutes on the fix branch, we hit a second, distinct manifestation of the same upstream bug (#5709). The detector does NOT fire on this variant. Caught it via Prometheus metrics:

| Metric | Value |
|--------|-------|
| `sync_downloads_in_flight` | **0** |
| `checkpoint_queued_slots` | 363 |
| `checkpoint_queued_continuous_height` | 1,732,117 |
| `checkpoint_queued_max_height` | 1,732,614 |
| `state_tip` | 1,732,115 |
| `sync_cancelled_verify_count` | 455 (climbing) |
| `sync_cancelled_download_count` | 45 |

State hadn't advanced for 10+ minutes. The checkpoint verifier has 363 blocks queued — contiguous from `state_tip+1` only up to `state_tip+2`, then a gap, then non-contiguous fragments out to `state_tip+499`. The syncer's `in_flight` is **zero** — it has stopped requesting the gap-filling block(s) entirely.

This is the same underlying bug — checkpoint verifier needs strict contiguity, one block missing parks the rest — but presents inversely from the symptom the current detector targets:

- **Original variant (detector catches):** in_flight high, queue saturated → reservation pressure
- **Second variant (detector misses):** in_flight = 0, queue gap-parked → request pressure has collapsed

The syncer apparently times out on the missing block and never re-issues a request for it. The 363 already-queued blocks above the gap are just dead weight.

## Suggested detector extension

Either an OR-condition or a second detector for the gap-parked case. The cheap signal: `checkpoint_queued_max_height - checkpoint_queued_continuous_height` (the "queue gap span") is **497** in our case. If that gap span stays high and `in_flight` is low and the tip is frozen for >stall_restart_timeout, restart from tip.

Pseudocode:
```
if state_tip_unchanged_for(stall_restart_timeout):
    if in_flight >= saturation_threshold:
        # current detector — saturation variant
        restart_sync()
    elif (queue_max_height - queue_continuous_height) >= gap_threshold:
        # NEW — gap-parked variant
        restart_sync()
```

The metric `checkpoint_queued_*` is already exported, so the data needed is already in the verifier. The detector just needs to read it.

## Repro environment
Reproducible: hit it consistently in the height 1.72-1.74M region (very early sandblasting block density). The state of the node is preserved if anyone wants to test detector candidates against the same database — happy to provide a snapshot.

## Operator-side workaround for now
External watchdog re-enabled at a 90s threshold (catches both variants). Bouncing the process when the gap span exceeds ~400 also works as a targeted manual intervention.

Big thanks for the fix branch and the original diagnosis — getting from "8-minute wedge cycles" to "one variant solved + one new variant identified with clear extension path" is real progress in a day.

**Comment by oxarbitrage** [comment]
@zmanian — our previous comment crossed with yours; I hadn't seen the branch when I posted it. We'll review `fix/5709-sync-stall-pr` before opening a new one.

**Comment by zmanian** [comment]
### A second manifestation (idle gap-park) — and why an internal restart doesn't clear it

Testing the fast-restart approach on a fork branch, we hit a related stall that a saturation-keyed detector misses. 2-minute window, 5s Prometheus scrape (mainnet, ~height 1.73M):

| metric | t=0–25s | t=25s (collapse) | t=25–80s |
|---|---|---|---|
| `sync_downloads_in_flight` | 270 | → **0** | 0 (held ~55s) |
| `checkpoint_queued_slots` | 250 | 250 | 250 |
| `checkpoint_queued_continuous_height` | 1,732,118 | 1,732,118 | 1,732,118 (frozen) |
| `checkpoint_queued_max_height` | 1,732,613 | 1,732,613 | 1,732,613 |
| `sync_cancelled_verify_count` | 0 | **250** | 250 (flat) |
| `zcash_chain_verified_block_total` | flat | flat | flat |

The checkpoint verifier is parked waiting for the block just above `continuous_height` (1,732,119), while holding 250 blocks scattered above the gap up to `max_height`. `cancel_verify` jumping to 250 then staying **flat** (not climbing) confirms this is a genuine idle park, not saturation churn — the node then sits at `in_flight=0` with the tip frozen until an external watchdog bounces the process. The collapse signature (all downloads **and** all queued verifies cancelled in one instant, with no stall-detection log) is a generic `sync error, restarting` (`cancel_all()` + the 67s `SYNC_RESTART_DELAY`).

Two things this surfaces, both directly relevant to a fast-restart fix:

**1. An internal restart doesn't clear the verifier's queue.** `cancel_all()` clears the *syncer's* in-flight downloads, but not `CheckpointVerifier`'s internal `queued` map — `reset_progress()` resets only the progress marker, not `queued`, and nothing else clears it on a syncer restart. So the re-walk from the state tip re-enters a verifier that still holds the **same stale gap-parked range** and re-stalls on the same missing block. The corroborating signal is operational: a full **process bounce recovers** (yields a burst), but an internal `cancel_all` restart does not — and the clean distinguishing factor is that a fresh process starts with an empty verifier queue. This suggests recovery needs a **verifier-queue reset back to the state tip**, not just re-issuing download requests faster.

**2. Detection should key off the verifier's gap state, not download-queue saturation.** The node spends most of the cycle at `in_flight=0` (in the restart delay), where any detector gated on download-queue saturation can't see the parked verifier. The verifier's `WaitingForBlocks` gap state + a frozen state tip is the signal that catches both the saturated and the idle manifestations.

Net: a faster restart alone may not help this case — it re-stalls on the persistent verifier queue. We're going to test a verifier-reset-on-stall variant on the node and will report whether it self-heals without a process bounce. Happy to share the full metric trace meanwhile.

**Comment by zmanian** [comment]
### Two experiments: ruled out queue persistence, found+fixed a restart doom-loop, isolated the remaining blocker

Update from testing on a node stuck at height 1,732,115 in the dense ~1.72-1.74M region. Two experiments since the last post.

**Experiment 1 — verifier-queue reset (ruled out).** I'd hypothesized recovery needed clearing the `CheckpointVerifier`'s stale queue on restart. Built and tested it: the reset works (queue clears, refills) but `verified_height` never advances across 5 cycles / 10 min. **So queue persistence is not the blocker** — correcting my earlier guess.

**Experiment 2 — don't restart the syncer on a single block download failure (the real find).** Diagnostic capture showed the smoking gun: a single `DownloadFailed { error: SharedPeerError(ConnectionClosed) }` at 15:02:02 → wholesale collapse at 15:02:04 (`cancelled.verify` = 471, exactly one `sync error, restarting`). `should_restart_sync` only continues on `DownloadFailed` when the error string contains `"NotFound"`; `ConnectionClosed` falls to the catch-all `_ => true` → full restart → `cancel_all()` discards the **entire** queued checkpoint range. In this dense region a contiguous range takes long enough to assemble that a transient peer drop almost always interrupts it first → a restart doom-loop where the node never commits.

Patch (drop the single failed block and continue, instead of restarting; systemic `NetworkServiceError`/`VerifierServiceError` still restart) demonstrably removes the doom-loop:

| | before (restart-on-error) | after (drop+continue) |
|---|---|---|
| wholesale cancellations | ~5 / 10 min | **1 / 10 min** |
| queue persistence | reset every ~90s | **persists 5+ min** |
| contiguous past tip | spikes then collapses | **4 blocks held (tip+1..+4)** |
| `in_flight=0` samples | 50+ | 4 (~20s) |

This aligns with the standing TODO on that line (#2908: handle by error type, not string-match).

**But a second blocker remains — and it's a contiguity gap, confirmed.** Even with the queue persisting, `verified_height` stays at 1,732,115 while `checkpoint_queued_continuous_height` reaches only 1,732,119. The decisive metric: `checkpoint_processing_next_height = 1,732,115` — equal to the verifier's anchor, which is the `WaitingForBlocks` signature (no checkpoint reachable within the contiguous run, so `target_checkpoint_height()` sets the gauge to `start_height`). So the verifier is correctly **holding** the 4 contiguous blocks (tip+1..+4) and waiting for **1,732,120** — not rejecting present blocks.

Meanwhile the syncer keeps downloading blocks *above* the gap — `max_height` 1,733,511, `queued_slots` climbing (431 → 604 and counting) — but **never re-fetches 1,732,120 itself**. So removing the doom-loop reveals the underlying defect cleanly: the syncer downloads forward past the unfillable gap (toward unbounded queue growth) and never re-requests the one block it's stuck on. That's the request side — `obtain_tips`/`extend_tips` (FindBlocks fanout + last-hash stripping) leaving a hole in the hash sequence — i.e. the "why does a given block go missing" question, now localized to the tip-walk rather than download or verify.

Happy to share: the `should_restart_sync` patch branch (`zmanian/zebra@exp/5709-no-restart-on-download-fail`), the full metric traces, and a **snapshot of the stuck state DB** if you want to reproduce the tip-walk gap against a fixed database. The restart-on-error fix looks worth landing on its own; the tip-walk gap is the part that overlaps your "less settled" question.

**Comment by zmanian** [comment]
## Summary

We reproduced the long-standing initial-sync stall on a cloud node and traced it to a single root cause: **stale "missing inventory" marks make individual blocks un-routable, which freezes contiguous checkpoint verification at one block until the marks rotate out (~1–1.5 min) — or indefinitely if they keep getting refreshed.**

Along the way we found and fixed five distinct defects, from "the stall is completely silent" to "the block download never actually fails, it's just un-routable." The node now advances on its own, but the central fix touches a deliberate DoS-protection tradeoff in `route_inv`, so we'd like the team's view before proposing it for merge.

Everything below is on branches on `zmanian/zebra`; happy to open PRs for any piece, and we can provide a DB snapshot positioned right before a stall for deterministic repro.

## Root cause

The peer set keeps an **inventory registry**: when a peer returns `NotFound` for a hash — or a connection drops mid-request — that peer is marked "missing" that hash, and `route_inv` then avoids peers marked missing (to avoid wasting bandwidth re-querying inventory a peer said it lacks). Correct for transactions; a trap for **blocks during initial sync**:

1. The block provably exists on the canonical chain. A `NotFound` is almost always transient — peer load, peer still syncing to that height, lag, or a dropped connection.
2. The mark persists for up to **2× `INVENTORY_ROTATION_INTERVAL`** (`current` + `prev` maps, rotated every 53s → entries live 53–106s).
3. Under load the *same* in-demand block gets a `NotFound` from peer after peer until **every** ready peer is marked missing it. `route_inv` then returns a synthetic `NotFoundRegistry` immediately.
4. The checkpoint verifier commits blocks only as a **contiguous run**. One un-routable block at height *N* parks every downloaded block above *N*. Frontier sits at *N−1*, CPU ~0%, disk idle, until the marks rotate out.

So the visible symptom — stuck at a fixed height, no CPU, no disk, peers connected — is the registry de-routing one block, not a hang, fork, or bandwidth.

## Evidence

From instrumented runs on the stuck node:

- **~59% of block requests returned `NotFound`** (588 of 997 in one window), split between `NotFoundResponse` (peer said no) and `NotFoundRegistry` (all peers marked missing).
- **The stuck block was tiny** — e.g. height 1,732,225 was 66 KB / 3 transactions. Rules out throughput/bandwidth and TCP slow-start.
- **The tip matched the canonical chain byte-for-byte** at the stall height. Rules out a fork.
- **Stall duration tracked the rotation window** (~60–106s per poisoned block).
- With the fixes below the node **advances on its own**, but ~one poisoned block per ~1.5 min — exactly the de-route window. That residual is what the relief (defect 5) targets.

Two surfaces of the same cause: `in_flight` pinned high (~999), and `in_flight = 0` idle gap-park (frontier block fails fast via `NotFoundRegistry`).

## Defects found, and fixes

All on `exp/5709-retry-missing-blocks` (cumulative). Defect 1 is also a standalone, conservative, config-gated branch (`fix/5709-sync-stall-pr`).

1. **The stall is completely silent.** Fix: `CheckpointVerifier` publishes a `watch` "contiguity-gap" signal (the height it's waiting to extend past); the syncer logs a WARN when the gap persists and — gated behind `sync.stall_restart_timeout` (default 90s, `"0s"` disables) — fast-restarts from the current tip. Observability + safety net; no change to verification. *(standalone PR branch)*
2. **A single block download failure restarted the whole syncer.** `should_restart_sync` treated one `DownloadFailed` as a reason to tear down all in-flight downloads. Fix: drop the failed block and continue.
3. **Dropped connections poisoned inventory routing.** `MissingInventoryCollector` marked a peer missing on *any* error, incl. `ConnectionClosed`. Fix: don't record missing inventory for `ConnectionClosed` (a genuine `NotFoundResponse` still records).
4. **Block download could wait unbounded on a concurrency permit** (acquired in `poll_ready` before the per-request timeout starts). Fix: outer `BLOCK_DOWNLOAD_QUEUE_TIMEOUT` (90s) bounding total time including the permit wait.
5. **(Core relief.) Blocks go un-routable when all peers are marked missing.** Fix (experimental, **needs the team's call**): in `route_inv`, when every ready peer is marked missing a **block** hash, retry one ready peer anyway instead of `NotFoundRegistry`. **Transaction inventory unchanged** — still fails, preserving DoS protection. New tests: `peer_set_route_inv_all_missing_block_retries`, `peer_set_route_inv_all_missing_tx_fails`.

## The tradeoff in defect 5

The existing `route_inv` behavior is intentional (don't re-query peers for inventory they've said they lack). Our relief deliberately re-queries for blocks in the exhaustion case, scoped to blocks only, one peer per request. Options the team might prefer:

- **(a)** Shorten `INVENTORY_ROTATION_INTERVAL` (53s → ~15–20s) so stale marks age out faster — simpler, but affects all inventory tracking.
- **(b)** Don't record missing inventory for `NotFoundResponse` on **block** requests at all — fixes at source, loses block availability tracking.
- **(c)** Retry-on-exhaustion for blocks (this branch) — keeps availability tracking + P2C preference; only changes the all-missing fallback.

We went with (c) as the most surgical, but this is the maintainers' tradeoff and we're happy to switch.

## Branches (on `zmanian/zebra`)

- **`fix/5709-sync-stall-pr`** — defect 1 only: gap signal + WARN + config-gated fast-restart. No change to verification or routing. Ready to PR on its own.
- **`exp/5709-retry-missing-blocks`** — full stack (defects 1–5), incl. the relief. Experimental; depends on the team's decision about defect 5.

## What would help

1. A read on defect 5: is (c) acceptable, or (a)/(b)/something else?
2. Whether to land defect 1 (the config-gated net) on its own now.
3. We can provide a DB snapshot positioned just before a stall for deterministic repro.

Cross-ref: env-var collection config bug filed while debugging — #10658.

**Comment by gustavovalverde** [comment]
Hey @zmanian, your root cause matches what we have been tracking on our side, and the "parked in verify, not download" evidence aligns closely with our diagnosis. Take this as independent confirmation that you might be on the right track:

**The root cause appears to involve three interconnected issues:** Firstly, the missing-inventory registry is corrupted by transient failures, such as timeouts, dropped connections, or brief overloads, which are mistakenly recorded as "this peer lacks this block." This is the trap you identified. Secondly, the syncer interprets a single download failure as a cue to tear down and restart the entire process, discarding progress over a problematic block. Thirdly, the checkpoint verifier only accepts strictly contiguous sequences, so encountering an un-routable block at height N stalls all subsequent blocks. This results in a frozen tip, with idle CPU and connected peers, precisely as you observed.

How we've been testing it, in dependency order:

1. **Observability.** Uses typed restart and error counters, along with a checkpoint contiguity-gap gauge indicating the height the verifier is monitoring. It shares the same signal as your verifier watch; we interpret it as a metric, not a restart trigger.
2. **Stop poisoning, recover instead of restarting.** Only log missing inventory on explicit `notfound` responses, not on timeouts, drops, or overloads (your defect 3). Replace the string-matched `should_restart_sync` with typed recovery that requeues the failed block at the same position, inserting missing predecessors before later hashes, instead of tearing down the run (your defect 2).
3. **Source-aware routing.** Route each block's `getdata` request to the peer that announced it, and enforce the contiguous-next prefix onto reliable routing so the verifier's next block is always received. The frontier beyond it remains source-aware.
4. **Peer quality.** Reduce priority for genesis-responders—peers that answer tip discovery from genesis while claiming to be ahead—so they stop winning slots and causing the `notfound` storm.

Regarding your defect 5, your understanding is correct. We tested avoiding the exhaustion scenario by using source-awareness and a no-poison policy instead of relaxing the registry. However, a true `notfound` from a peer still syncing to that height can still build up on the forced-reliable path, making that residual situation valid. Among your options, block-scoped retry-on-exhaustion (c) integrates most cleanly with source-aware routing: it only alters the all-missing fallback, ensures transactions still fail, and maintains DoS protection. I recommend leaning towards this approach, focusing solely on blocks.

**Comment by zmanian** [comment]
## Update from live testing on the full fix stack

We ran the full stack (`exp/5709-retry-missing-blocks`) on the stuck cloud node. Two findings, one of which **directly affects the team's in-flight fast-restart fix.**

## Finding 1 — the relief works; what it exposed underneath

The fixes did their job: the node got **past** the request-layer wedge (the `NotFoundRegistry` inventory-poisoning that caused the silent ~0%-CPU stalls) and is now making forward progress (~94 blocks / 30 min, committing in checkpoint-range bursts).

What it ran into next is a **different, CPU-bound** bottleneck, not a deadlock. A `perf` profile of the pegged core during a "stall":

```
60.00%  bls12_381::scalar::Scalar::square
13.06%  ff::helpers::sqrt_tonelli_shanks
12.86%  bls12_381::scalar::Scalar::mul
 3.88%  bls12_381::scalar::Scalar::invert
 2.24%  ff::Field::pow_vartime
```

This is `bls12_381` **scalar-field** arithmetic dominated by Tonelli–Shanks square roots — i.e. **jubjub point decompression** (jubjub's base field is BLS12-381's scalar field). That's **Sapling** data being parsed/loaded, not Orchard (which is Pasta/Halo2 and would show `pasta_curves` frames), and not Groth16 **proof verification** (which would show pairing frames — `multi_miller_loop`, `Fp12`). So the checkpoint verifier is correctly *not* verifying proofs; the cost is deserializing/decompressing Sapling-heavy sandblasting-era blocks. It's slow but bounded (the band is finite), and it progresses on its own.

This isn't a Zebra bug so much as the well-known sandblasting IBD cost surfacing once the request-layer wedge is gone. (A `perf record -g` call-graph would pin whether the decompression is in block deserialization vs. note-commitment-tree building; happy to share one.)

## Finding 2 — the fast-restart doom-loops in sandblasting (affects the team's fix)

This is the important one. The contiguity-gap **fast-restart** (our defect-1 fix, and the same approach the team is shipping) **doom-loops through the sandblasting region.**

Mechanism: the detector restarts when the contiguity frontier hasn't advanced within `stall_restart_timeout`. But a single sandblasting-era checkpoint range legitimately takes **minutes** to download + verify, so the frontier is static *while the verifier is still working*. The detector can't tell "frontier static because busy on a long range" from "frontier static because wedged" — both look identical from the frontier signal. So at the 90s default it false-fires, cancels all in-flight downloads, re-downloads, and the verifier starts the range over — never finishing. On the live node this showed up as **~4,199 wholesale download cancellations**; only `stall_restart_timeout = "0s"` (disable) let it make progress.

The original `BLOCK_VERIFY_TIMEOUT = 8 min` backstop has the same flaw, just triggers less often. **A bigger constant isn't the fix** — the detector needs a *progress-based liveness signal*, not a frontier-advance timer.

Our fix (on the branch): the `CheckpointVerifier` publishes a monotonic **liveness counter** alongside the gap signal (folded into the existing watch channel). It ticks on every `queue_block` (a downloaded, deserialized, checked block) and per verified block in `process_checkpoint_range`. The syncer snapshots it when arming the deadline; `detect_gap_stall` only declares a stall if the verifier did **no work** since — so a busy-but-slow verifier is left alone, while the original ~0%-CPU inventory wedge still restarts. Tests: `no_stall_when_verifier_made_progress` (unit) and `syncer_does_not_restart_when_verifier_is_busy` (paused-time integration).

**If the team's fast-restart lands without an equivalent liveness gate, it will doom-loop every node through sandblasting.** That's the thing worth catching before it ships.

## Suggested merge tiering (by evidence)

- **Solid:** no-restart-on-single-`DownloadFailed` (evidenced — it's what let the verifier grind uninterrupted); no-poison-on-`ConnectionClosed`; the gap-watch signal + WARN observability.
- **Needs the liveness gate:** the fast-restart action itself — safe only with the progress-based guard above.
- **Experimental / your design call:** the `route_inv` retry-on-exhaustion relief for blocks (the request-layer fix), and the bound-permit-wait timeout.

## Branches (on `zmanian/zebra`)

- `exp/5709-retry-missing-blocks` — full stack incl. the relief and the new liveness-gated fast-restart.
- `fix/5709-sync-stall-pr` — the conservative gap-signal + config-gated restart, defect-1 only.

Status: the node is progressing through the sandblasting band (days at the current rate; a CPU-native build for faster field arithmetic is being tried). The validation that closes this out is reaching post-sandblasting heights and confirming `in_flight` no longer saturates into the original wedge — we'll report when it gets there. DB snapshot for deterministic repro still on offer.

**Comment by zmanian** [comment]
Thanks @gustavovalverde — that lines up, and your clarifications reframe our last comment (our posts crossed).

**Gap signal as a metric, not a trigger:** that's the distinction I missed. Our "fast-restart doom-loop" finding is really evidence *for* your choice — we wired the gap to a fast-restart and it doom-looped through sandblasting (a range legitimately takes minutes, the frontier sits static, the timer false-fires and cancels in-progress work). So treat it as corroboration that restart-on-gap is the wrong recovery, not a warning about your design. The monotonic verifier-liveness counter we added (ticks per queued/verified block) is a clean "working vs wedged" *metric* if useful.

**Recovery:** your typed recovery that requeues the failed block (+ missing predecessors) is more complete than our drop-and-continue — ours leaves the tip-walk hole we flagged, which is why we needed the route_inv relief to compensate. Requeue + source-aware routing fixes it at the source; our (c) is the residual fallback you described.

**Defect 5:** thanks for the steer — we'll keep block-scoped retry-on-exhaustion as-is (transactions still fail, DoS protection intact) so it slots under source-aware routing.

The observability half of `fix/5709-sync-stall-pr` (gap gauge + WARN) matches your #1; the config-gated restart half is the part our own doom-loop finding argues against, so don't read that as a recommendation. Additive pieces from us: (c), the no-poison policy, and the paused-time tests + a stuck-state DB snapshot for repro — happy to adapt to your branch structure.

**Comment by zmanian** [comment]
Status update from running the experimental fix stack for ~3 days on mainnet initial sync. Picking up from the request-trace analysis I posted earlier.

## TL;DR

The combination of fixes on [zmanian/zebra@exp/5709-retry-missing-blocks](https://github.com/zmanian/zebra/tree/exp/5709-retry-missing-blocks) (commit `c048aca18`) unstuck the original wedges. The node sustained ~217K blocks of progress (1,732,224 → 1,949,918 so far) across the sandblasting region and is now grinding toward tip at meaningful rates. The "silent multi-minute wedge" pattern is gone. What remains is genuine verification cost.

## What worked

Running on commit `c048aca18` with:
- `ZEBRA_SYNC__CHECKPOINT_VERIFY_CONCURRENCY_LIMIT=400` (raised to 1000 once past sandblasting)
- `ZEBRA_SYNC__DOWNLOAD_CONCURRENCY_LIMIT=150` (250 once past sandblasting)
- `ZEBRA_SYNC__STALL_RESTART_TIMEOUT=90s` (default — the liveness gate added in this branch handles it correctly)
- TOML config pinning 40 known-good peers + DNS seeds
- `ZEBRA_NETWORK__EXTERNAL_ADDR` set to our public IP
- Host sysctl `net.ipv4.tcp_slow_start_after_idle=0`
- Binary built with `RUSTFLAGS="-C target-cpu=native"` (host is c4-standard-24 with AVX-512-IFMA)

What's specifically *not* present anymore:
- No silent multi-minute wedges where `in_flight=999` and the syncer task goes quiet
- No `cancelV` doom-loops where one transient peer error nukes 400+ queued blocks
- No need for an external process-bouncing watchdog (we disabled ours; the in-process liveness gate suffices)

## Observed rates by region

| Block region | Sustained rate | Notes |
|---|---|---|
| 1.43-1.72M (early Sapling) | ~67-150K/hr (burst) | Where the original wedges kept happening |
| 1.72-1.85M (heavy sandblasting) | 800-2,500/hr | Genuine Sapling/Orchard verification cost |
| 1.85-1.90M (late sandblasting) | 5,000-7,500/hr | Still expensive blocks, but moving |
| 1.90M+ (post-sandblasting) | 12,000-21,000/hr | Verification gets cheap again |

The sandblasting cost is real and unrelated to #5709 — it's `bls12_381::scalar::Scalar::square` + `sqrt_tonelli_shanks` dominating the perf top frames at 60-80% combined.

## Operational notes for other operators landing here

1. **`stall_restart_timeout=0s` is no longer needed.** The liveness gate in `c048aca18` makes the default 90s safe. Earlier in our investigation we needed `0s` to avoid the doom-loop; that's an artifact of the older branches.

2. **`checkpoint_verify_concurrency_limit=400` during sandblasting, default (1000) after.** We saw a real regression when raising to 1000 mid-sandblasting (~1.85M) — slower rate, higher `cancelV` — but the same setting works fine past ~1.90M. The bug it was mitigating is more pronounced when block verify cost is high.

3. **External watchdogs do more harm than good now.** A time-based "no progress in N seconds → process bounce" loop misinterprets legitimately-slow sandblasting blocks (5-8 min single-block verify) as stalls. Either set N > 15 min or remove the watchdog entirely. The in-process liveness gate has the progress signal it needs without watchdog help.

4. **`target-cpu=native` is worth it.** On a CPU with AVX-512-IFMA the BLS12-381 hot loops parallelize meaningfully better; verifier becomes ~50% faster through sandblasting blocks (rayon threads appear in perf top alongside the tokio runtime workers). Generic x86_64 builds leave throughput on the table.

5. **The `sync_cancelled_verify_count` rate is a noisier-than-expected signal.** Even on a well-behaved sync we see 100-400/min cancellations — they don't block progress because the bursts after each cancellation are large enough to net-positive. Watch verified-height advance over windows of minutes, not the cancellation counter.

## What I'd suggest for upstream

- The `c048aca18` change set (liveness gate, no-restart-on-transient-error, no-poison-on-transient-error, retry-missing-blocks, bound-permit-wait, verifier-reset) seems to net to a real fix. I don't have the granular A/B data to credit individual changes — would defer to maintainers on which to land and which to merge into a smaller diff.

- The diagnostic instrumentation in `exp/5709-instrument-request` (dbg5709 lifecycle logs + `checkpoint_processing_next_height` / `checkpoint_queued_continuous_height` / `checkpoint_queued_max_height` metrics) was load-bearing for the debug. Worth landing those metrics independently of the fix — they would have made this whole investigation 10x faster if they existed before #5709 was filed.

- The `sandblasting verifier cost` and the `#5709 stall` are two distinct problems that present similarly (long time-since-last-state-block). They're easy to confuse in the field. A `checkpoint_verifying_height` gauge that says "the verifier is currently chewing on block N" would let operators tell them apart at a glance — distinct from `checkpoint_processing_next_height` (which is the next-to-process, not the currently-grinding one).

Happy to provide more granular per-cycle data, the stuck-state DB snapshot we captured earlier, or run additional targeted experiments if useful.

Big thanks to everyone who's worked this — the difference between "8-minute silent wedge every few minutes" and "21K blocks/hr post-sandblasting" is operationally enormous.

**Comment by oxarbitrage** [comment]
Thanks for the detailed report. We're trying to isolate the smallest possible fix and land other hardening as follow-ups.

We're targeting a fix that only addresses the FindBlocks availability registration gap (so `route_inv` routes to the right peer) and re-requesting dropped NotFound blocks — 2 files, ~100 lines. We're testing it against our fleet on 1 node for now and the early results have been clean (0 Elapsed/Duplicate errors vs every other stock node showing them), but the test just started. We should have more data tomorrow.

If that's not enough on its own, we'll look at pulling in changes from your stack, from other approaches we have in review, or new ideas as we learn more. You mentioned you can't isolate individual changes — that's what we're trying to work out from our side. More info on this issue soon.
